File: epi_judge_java/epi/ReverseLinkedListIterative.java

package epi;

public class ReverseLinkedListIterative {

  public static ListNode<Integer> reverseLinkedList(ListNode<Integer> head) {
    ListNode<Integer> prev = null, curr = head;
    while (curr != null) {
      ListNode<Integer> temp = curr.next;
      curr.next = prev;
      prev = curr;
      curr = temp;
    }
    return prev;
  }

  private static <T> void print(ListNode<T> head) {
    if (head != null) {
      System.out.println("(" + head.data + ")");
      print(head.next);
    }
  }

  public static void main(String[] args) {
    ListNode<Integer> l1 = new ListNode<>(1, null);
    ListNode<Integer> l2 = new ListNode<>(2, null);
    l1.next = l2;
    ListNode<Integer> l3 = new ListNode<>(3, null);
    l2.next = l3;

    System.out.println("before reverse");
    print(l1);
    ListNode<Integer> newhead = reverseLinkedList(l1);
    System.out.println("\nafter reverse");
    print(newhead);
    newhead = reverseLinkedList(newhead);
    System.out.println("\nafter another reverse");
    print(newhead);
  }
}

File: epi_judge_python_solutions/reverse_linked_list_iterative.py
from list_node import ListNode


def reverse_linked_list(head):
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev


def print_list(head):
    if head:
        print('(', head.data, ')', sep='')
        print_list(head.next)


def main():
    L1 = ListNode(1, ListNode(2, ListNode(3, None)))

    print('before reverse')
    print_list(L1)
    newhead = reverse_linked_list(L1)
    print('\nafter reverse')
    assert newhead.data == 3 and newhead.next.data == 2 and newhead.next.next.data == 1
    print_list(newhead)
    newhead = reverse_linked_list(newhead)
    assert newhead.data == 1 and newhead.next.data == 2 and newhead.next.next.data == 3
    print('\nafter another reverse')
    print_list(newhead)


if __name__ == '__main__':
    main()
